core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala

```diff
@@ -236,6 +236,7 @@ import scala.util.control.NonFatal
 
   /** ActorRefs of all stages that sent subscriptions requests or `RegisterSubStage` to this actor (removed on their termination). */
   private var stageActors = Set.empty[ActorRef]
+  private var stageActorsMap = Map.empty[TopicPartition, ActorRef]
```

Comment:
I wonder if we can somehow consolidate with `stageActors` here, but I see that `stageActors` is also referenced for single sources.

Reply:
Do we need to manage `stageActorsMap` in the `stopping` receive block too?

Reply:
@seglo I have consolidated stageActors and stageActorsMap and also extended the fix for single sources. Please review. Here are the test results on my dev machine

```
branch::master
[info] Passed: Total 322, Failed 0, Errors 0, Passed 322                        
[success] Total time: 1192 s (19:52), completed Nov 26, 2020 12:36:18 AM        
                                                                                
branch::message-loss-test-case-and-fix                                                
[info] Passed: Total 323, Failed 0, Errors 0, Passed 323                        
[success] Total time: 1073 s (17:53), completed Nov 26, 2020 2:41:13 AM  
```

There is some improvement in total test time as well (19.minutes vs. 17.minutes), so I am hoping the fix doesn't have any performance impact

Reply:
> There is some improvement in total test time as well (19.minutes vs. 17.minutes), so I am hoping the fix doesn't have any performance impact.

That's interesting! I wonder if it was a fluke? I'll review this more thoroughly tomorrow.

Reply:
Sure. I have also added you to this repo, in case you want to push any changes to this branch.